colbuilder: fix recently introduced bug in an edge case

yuzefovich opened the pull request:
Our typing system is pretty broken, especially when it comes to the
integers of non-default width. The row-execution processors don't
respect the width of integers and always operate on INT8; however, the
vectorized engine respects the widths. We work around this complication
by having to plan casts occasionally.

In eb24f95 we introduced a subtle bug
when a wrapped noop processor is planned by updating `ColumnTypes` to
the desired once assuming that the noop processor always is able to
satisfy the desired schema. But this isn't true for integers of
different widths. Before that change we would correct plan vectorized
casts to get integers of the desired widths, but that commit overwrote
the actual schema with the desired one, so we no longer knew we needed
a cast. This is now fixed.

Fixes: #66306.

Release note (bug fix): CockroachDB could previously return an internal
error under rare circumstances when a query involving integers of widths
2 or 4 was executed in the distributed manner via the vectorized engine.
The bug is a regression in 21.1.2 release (which is the only stable
release containing this bug).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66346)
<!-- Reviewable:end -->

yuzefovich committed:
colbuilder: fix recently introduced bug in an edge case

Our typing system is pretty broken, especially when it comes to the
integers of non-default width. The row-execution processors don't
respect the width of integers and always operate on INT8; however, the
vectorized engine respects the widths. We work around this complication
by having to plan casts occasionally.

In eb24f9578439ef6fbf5a3cc36fd3ee4a37611043 we introduced a subtle bug
when a wrapped noop processor is planned by updating `ColumnTypes` to
the desired once assuming that the noop processor always is able to
satisfy the desired schema. But this isn't true for integers of
different widths. Before that change we would correct plan vectorized
casts to get integers of the desired widths, but that commit overwrote
the actual schema with the desired one, so we no longer knew we needed
a cast. This is now fixed.

Release note (bug fix): CockroachDB could previously return an internal
error under rare circumstances when a query involving integers of widths
2 or 4 was executed in the distributed manner via the vectorized engine.
The bug is a regression in 21.1.2 release (which is the only stable
release containing this bug).

yuzefovich commented:
I intend to backport only the first commit.

I originally misdiagnosed the issue and added the second commit (which taped over the regression), but I think the second commit is sane, so I'm keeping it to merge only on master.

yuzefovich commented:
To be more explicit, the bug in eb24f9578439ef6fbf5a3cc36fd3ee4a37611043 is here:
```
@@ -1529,10 +1546,14 @@ func (r opResult) wrapPostProcessSpec(
        }
        inputToMaterializer := colexecargs.OpWithMetaInfo{Root: r.Root}
        takeOverMetaInfo(&inputToMaterializer, args.Inputs)
-       return r.createAndWrapRowSource(
+       if err := r.createAndWrapRowSource(
                ctx, flowCtx, args, []colexecargs.OpWithMetaInfo{inputToMaterializer},
                [][]*types.T{r.ColumnTypes}, noopSpec, factory, causeToWrap,
-       )
+       ); err != nil {
+               return err
+       }
+       r.ColumnTypes = resultTypes
+       return nil
 }
```
`createAndWrapRowSource` updates `r.ColumnTypes` to the actual types, but because of the bug we would set `ColumnTypes` to the desired schema (resulting in a mismatch between actual and desired, and not giving us an opportunity to plan a cast later).

yuzefovich commented:
Decided to remove the second commit altogether (for now) since there were some failures.

yuzefovich commented:
Friendly ping @michae2 - this bug is being hit often by the sqlsmith roachtest.

jordanlewis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66346)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 491 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66346#-McGeUOS3MWGeLU6tBeu:-McGeUOS3MWGeLU6tBev:b-b3kzyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dbf3170153b3a0c4cf442487ced546489ead100/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L491)):*
> ```
> # width of an integer column during a cast and the vectorized engine not
> # performing the cast to the integer of the desired width (#66306).
> statement ok
> ```

nit: why not make query II?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66346#-:-McGpo5UCh6GBTJl-PLm:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[pkg/sql/logictest/testdata/logic_test/dist_vectorize, line 491 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66346#-McGeUOS3MWGeLU6tBeu:-McGpi940cYlP74hsutP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dbf3170153b3a0c4cf442487ced546489ead100/pkg/sql/logictest/testdata/logic_test/dist_vectorize#L491)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: why not make query II?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Oh, actually eb24f95 didn't make it into 21.1.2 release, so there should be no stable release with this bug. I'll ping the docs people.
